### ISSUE TITLE
Regent: bug in index space launched fills of arrays

### ISSUE THREAD

**Issue by mariodirenzo** [StanfordLegion/legion#694]
The following script
```
import "regent"

local C = regentlib.c

struct Point {
   v1  : double;
   v2  : double[3];
}

task main()
   var is = ispace(int1d, 10)
   var r  = region(is, Point)
   var tiles = ispace(int1d, 2)
   var p_r = partition(equal, r, tiles);

   -- This works
   __demand(__index_launch)
   for c in tiles do
      fill((p_r[c]).v1, 0.0)
   end

   -- This works
   var v = array(0.0, 0.0, 0.0)
   __demand(__index_launch)
   for c in tiles do
      fill((p_r[c]).v2, v)
   end

   -- This one fails with:
   --test_fill.rg:34: loop optimization failed: fill value is not provably invariant
   --      fill((p_r[c]).v2, array(0.0, 0.0, 0.0))
   __demand(__index_launch)
   for c in tiles do
      fill((p_r[c]).v2, array(0.0, 0.0, 0.0))
   end
end

regentlib.start(main)
```
fails with the error
```
loop optimization failed: fill value is not provably invariant
      fill((p_r[c]).v2, array(0.0, 0.0, 0.0))
```

The issue seems to be related to having the definition of the constant array inside the index space launch

**Comment by elliottslaughter** [comment]
It's probably interpreting `array` as a function; if you instead write something like `[terralib.constant(`array(0.0, 0.0, 0.0))]` this might just work.

**Comment by mariodirenzo** [comment]
Yes, `[terralib.constant(`array(0.0, 0.0, 0.0))]` works.
I guess that it is not different from defining the variable before the loop, right?

**Comment by elliottslaughter** [comment]
Yes, that's pretty much right, though technically you can make a variable with a dynamic value, whereas `terralib.constant` has to be a compile-time constant.

I think the issue is the compiler doesn't know enough about `array` to be able to see it as a constant. This is probably going to be tricky to fix since `array` is a Terra primitive not a Regent one. I guess we could literally run `terralib.constant` on it to see if an expression can be constant folded. Otherwise we're talking essentially about hard-coding the semantics of Terra primitives into Regent, which seems undesirable even if it could be made to work.

**Comment by mariodirenzo** [comment]
ok, I'll use this workaround. Thank you.
Should we keep the issue open for reference?

**Comment by elliottslaughter** [comment]
Let's keep it. It's plausible we can fix this, though there's a risk that what I proposed above might slow down the compiler.
